### PR TITLE
Refactor `makeRequest`, `makeXMLHttpRequest` and `makeFetchRequest` request to be promise-based

### PR DESCRIPTION
**Pull request by smellyshovel**
Part 3 of series of refactoring PRs (splitted up https://github.com/maplibre/maplibre-gl-js/pull/2038).

1. Changes implementations of `makeRequest`, `makeXMLHttpRequest` and `makeFetchRequest` to be promise-based. Updates respectively the consumers of these functions (e.g. `getJSON`), but **does not change signatures of the consumer functions**.
2. Updates the `RequestParameters` type to be based on the built-in `RequestInit` type.
3. Introduces a couple of new types related to HTTP requests and responses.
4. Removes the unused `postData` function.
5. Updates ajax's unit tests. Other (missing) tests will come later with updated implementations for other methods.
6. Removes a special logic for making network requests from web workers: Fetch API and XHR are both available there nowadays.

### CODE REVIEW

File: src/style/style.ts [hunk 1]
```diff
@@ -1513,7 +1513,15 @@ class Style extends Evented {
     }
 
     getResource(mapId: string, params: RequestParameters, callback: ResponseCallback<any>): Cancelable {
-        return makeRequest(params, callback);
+        const request = makeRequest(params);
+
+        request.response
+            .then(response => callback(null, response.data))
+            .catch(err => {
+                if (err.message !== 'aborted') callback(err);
```
Comment (drwestco) [inline src/style/style.ts:1521]: The error comes from `AbortController`? Looks like the safer way to check the error type is by `name` or `code`. https://stackoverflow.com/a/71670006

File: src/util/ajax.ts [hunk 2]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+export type RequestParameters = RequestInit & { url: string; collectResourceTiming?: boolean };
+
+/**
+ * A type that tells the `makeRequest` which modifications to apply to the request before making it and how to treat
+ * response of the request after it's loaded based on the type of the raw data being loaded.
+ *
+ * @enum {RequestDataType}
+ */
+export const enum RequestDataType {
+    'string',
```
Comment (drwestco) [inline src/util/ajax.ts:207]: This is odd syntax for declaring enum members. It's actually creating a numeric-valued enum instead of string, as you might expect at first read.

File: src/util/ajax.ts [hunk 3]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+/**
+ * A generic type that represents a MapLibre asynchronous cancelable HTTP request.
+ *
+ * Represented by an object containing 2 fields:
+ *  - `response`: a `Promise` that (possibly) resolves with the request's result
+ *  - `cancel`: a function to cancel the request
+ *
+ * @typedef {Request}
+ */
+export type Request<T> = {response: Promise<T>} & Cancelable;
```
Comment (drwestco) [inline src/util/ajax.ts:221]: Why isn't `response` typed as `Promise<Response<T>>`? Could save some repetitive code elsewhere, since the request functions all return `Request<Response<T>>`. Or is there a need for a request to ever return a non-Response-wrapped value?

File: src/util/ajax.ts [hunk 4]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+    'string',
+    'JSON',
+    'ArrayBuffer'
+}
+
+/**
+ * A generic type that represents a MapLibre asynchronous cancelable HTTP request.
+ *
+ * Represented by an object containing 2 fields:
+ *  - `response`: a `Promise` that (possibly) resolves with the request's result
```
Comment (drwestco) [inline src/util/ajax.ts:216]: The `(possibly)` hedge just makes whoever uses this type ask more questions. Does the Promise ever get resolved with something other than the result? Or does it get rejected in those cases? Or what?

File: src/util/ajax.ts [hunk 5]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+export function makeRequest<T>(requestParameters: RequestParameters, requestDataType?: RequestDataType): Request<Response<T>> {
+    /*
+        See https://github.com/maplibre/maplibre-gl-js/discussions/2004
+
+        TL;DR: for the time being, it's still impossible to completely give up on using the XMLHttpRequest API. But
+        that's a point to reconsider in the (hopefully near) future
+     */
+
+    // if the url uses some custom protocol. E.g. "custom://..."
+    if (/:\/\//.test(requestParameters.url) && !(/^https?:|^file:/.test(requestParameters.url))) {
```
Comment (drwestco) [inline src/util/ajax.ts:261]: Good opportunity for cleanup here. Between this line, line 272, and line 281, this function has three different ways of checking the protocol.

File: src/util/ajax.ts [hunk 6]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+
+            return action(requestParameters, requestDataType);
+        }
+    }
+
+    // if there's no protocol at all or the protocol is not `file://` (in comparison with the `if` block above, it can
+    // now be `http[s]://`). E.g. "https://..." or "foo"
+    if (!requestParameters.url.startsWith('file://')) {
+        // and if Fetch API is supported by the target environment
+        if (fetch && Request && AbortController && Object.prototype.hasOwnProperty.call(Request.prototype, 'signal')) {
```
Comment (drwestco) [inline src/util/ajax.ts:283]: Why are these same checks not done before using `helper.makeFetchRequest` up on line 273?

File: package.json [hunk 7]
```diff
@@ -31,6 +31,7 @@
     "geojson-vt": "^3.2.1",
     "gl-matrix": "^3.4.3",
     "global-prefix": "^3.0.0",
+    "jest-fetch-mock": "^3.0.3",
```
Comment (HarelM) [inline package.json:34]: We have nise amd sinon, aren't they enough?

File: src/util/ajax.ts [hunk 8]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+    }
+
+    return {
+        response: (async (): Promise<Response<T>> => {
+            const response = await fetch(request);
+
+            if (abortController.signal.aborted) throw new Error('aborted');
+
+            if (response.ok) {
+                const data: T = await (requestDataType === RequestDataType.ArrayBuffer ? response.arrayBuffer() : requestDataType === RequestDataType.JSON ? response.json() : response.text());
```
Comment (drwestco) [inline src/util/ajax.ts:354]: Is lint working correctly? Or are nested ternary statements a-ok for this project?

File: src/util/ajax.ts [hunk 9]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
     }
     return {cancel: () => {}};
 };
+
+// new impl.
```
Comment (HarelM) [inline src/util/ajax.ts:190]: No need for this, unless you plan to remove it in future PR.

File: src/util/ajax.ts (hunk 2 above)
Comment (HarelM) [reply src/util/ajax.ts:207]: I would consider using `export type RequestDataType = "string" | "JSON"...`

File: src/util/ajax.ts [hunk 10]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+        TL;DR: for the time being, it's still impossible to completely give up on using the XMLHttpRequest API. But
+        that's a point to reconsider in the (hopefully near) future
+     */
+
+    // if the url uses some custom protocol. E.g. "custom://..."
+    if (/:\/\//.test(requestParameters.url) && !(/^https?:|^file:/.test(requestParameters.url))) {
+        // and if the request is made from inside a worker
+        if (isWorker() && (self as any).worker && (self as any).worker.actor) {
+            // then ask the main thread to make the request from there
+            return (self as any).worker.actor.send('getResource', requestParameters, requestDataType);
```
Comment (HarelM) [inline src/util/ajax.ts:265]: is `getResource` still a function that can be called from a worker to the main thread?

File: src/util/ajax.ts (hunk 5 above)
Comment (HarelM) [reply src/util/ajax.ts:261]: Yup, that's true.

File: src/util/ajax.ts [hunk 11]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+
+    if (requestDataType === RequestDataType.ArrayBuffer) {
+        xhr.responseType = 'arraybuffer';
+    }
+
+    for (const k in requestParameters.headers) {
+        xhr.setRequestHeader(k, requestParameters.headers[k]);
+    }
+
+    if (requestDataType === RequestDataType.JSON) {
```
Comment (HarelM) [inline src/util/ajax.ts:398]: Feels like this should be "else if" since we checked arrayBuffer

File: src/util/ajax.ts [hunk 12]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+        xhr.responseType = 'text';
+        xhr.setRequestHeader('Accept', 'application/json');
+    }
+
+    xhr.withCredentials = requestParameters.credentials === 'include';
+
+    xhr.send(requestParameters.body?.toString());
+
+    return {
+        response: new Promise<Response<T>>((res, rej) => {
```
Comment (HarelM) [inline src/util/ajax.ts:408]: Use full names - reject, resolve.
Res can be request and resolve and thus confusing.

File: src/util/ajax.ts [hunk 13]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+                        }
+                    }
+
+                    res({
+                        data,
+                        cacheControl: xhr.getResponseHeader('Cache-Control'),
+                        expires: xhr.getResponseHeader('Expires')
+                    });
+                } else {
+                    // const body = new Blob([xhr.response], {type: xhr.getResponseHeader('Content-Type')});
```
Comment (HarelM) [inline src/util/ajax.ts:427]: Hu?

File: src/util/ajax.ts [hunk 14]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+        signal: abortController.signal
+    }));
+
+    if (requestDataType === RequestDataType.JSON) {
+        request.headers.set('Accept', 'application/json');
+    }
+
+    return {
+        response: (async (): Promise<Response<T>> => {
+            const response = await fetch(request);
```
Comment (HarelM) [inline src/util/ajax.ts:349]: Shouldn't there be a try-catch here? Or is the promise rejection intended to bubble up?

**Comment by HarelM** [comment]
Very nice work overall!! The comments made are mainly cosmetic.

File: src/util/ajax.ts (hunk 8 above)
Comment (HarelM) [reply src/util/ajax.ts:354]: We are generally ok with those, but I'm guessing this could be simplified maybe, not sure though.

Comment (zhangyiatmicrosoft) [reply src/util/ajax.ts:354]: Hard to read for the next person, and should you ever need set break point here, nested ternary is bad.
I'd suggest breaking it down to simple if statements. Minimizer will do its job and end up the same.

File: src/util/ajax.ts (hunk 3 above)
Comment (smellyshovel) [reply src/util/ajax.ts:221]: There's a need to be able to address the `Response` type by its own, without the context of `Request`. And while that is still achievable using something like `Awaited<Request["response"]>`, I think that it overcomplicates the code making it hard to read and understand.

Generally, I agree with you, but long story short there are implementation restrictions. I'll try to check it further more, but without any promises that this is gonna be changed.

File: src/util/ajax.ts (hunk 4 above)
Comment (smellyshovel) [reply src/util/ajax.ts:216]: Sometimes it just does not get resolved nor rejected at all. Namely, when the request was aborted. This is how it used to be done with callbacks. If there's an error, call the callback with `cb(error)`. If there's result, call with `cb(null, result)`, but if the request was aborted - don't call the callback at all.

I don't like the idea of keeping promises in the pending state without ever resolving or rejecting them, but:
1. Generally, it's safe to do. No memory leakage should be brought by such an action
2. Such an approach allows to avoid changing consumers' implementations (as well as thier related unit tests) that much
3. Later I want to refactor this to through an error when a request gets aborted. But, once again, as we try to keep PRs as small and atomic as possible, that should be done later separately.

As further refactored, this `(possibly)` statement will be removed. But for now it indicates what's really happening.

File: src/util/ajax.ts (hunk 6 above)
Comment (smellyshovel) [reply src/util/ajax.ts:283]: I don't understand the question. Could you please elaborate more on it?

File: src/util/ajax.ts [hunk 15]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+            xhr.onerror = () => rej(new Error('Failed to Fetch URL'));
+
+            xhr.onabort = () => rej(new Error('aborted'));
+        }),
+
+        cancel: () => xhr.abort()
+    };
+}
+
+export const helper = {
```
[the remark this answers is not in the document]
Comment (smellyshovel) [reply src/util/ajax.ts:441]: To be able to spy on these functions in unit tests. If you know a better way to do so - I'll be happy to find out.

File: src/util/ajax.ts [hunk 16]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+
+    if (requestDataType === RequestDataType.JSON) {
+        request.headers.set('Accept', 'application/json');
+    }
+
+    return {
+        response: (async (): Promise<Response<T>> => {
+            const response = await fetch(request);
+
+            if (abortController.signal.aborted) throw new Error('aborted');
```
[the remark this answers is not in the document]
Comment (smellyshovel) [reply src/util/ajax.ts:351]: Do you mean that I don't need that `if` check at all? The `Promise` would automatically get rejected for me in case the request gets canceled?

File: src/util/ajax.ts (hunk 9 above)
Comment (smellyshovel) [reply src/util/ajax.ts:190]: I do plan so. It's easier for me to navigate through the code having this around. I'll remove it as soon as I'm done with refactoring of the whole file, Ok?

File: src/util/ajax.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (HarelM) [reply src/util/ajax.ts:441]: You can listen to the other side using nise I think, splitting ajax to another helper to be able to test it is a bit too much I think.
ajax is the "data access layer" and the way to test it is to mock fetch and xhr I believe.

File: src/util/ajax.ts (hunk 9 above)
Comment (HarelM) [reply src/util/ajax.ts:190]: Sure, just wanted to make sure this wasn't a leftover...

File: src/util/ajax.ts (hunk 3 above)
Comment (drwestco) [reply src/util/ajax.ts:221]: I'm not saying get rid of the `Response<T>` type, just change the signature for Request:
```
export type Request<T> = { response: Promise<Response<T>> } & Cancelable;
```
Since that's what you're doing everywhere I can find, by having return values of type `Request<Response<T>>`.

File: src/util/ajax.ts (hunk 4 above)
Comment (drwestco) [reply src/util/ajax.ts:216]: > Sometimes it just does not get resolved nor rejected at all. Namely, when the request was aborted.

That deserves to be in the comment. Make it clear that the Promise either resolves with the result, rejects with an error code, or is left pending if abort is called.

File: src/util/ajax.ts (hunk 6 above)
Comment (drwestco) [reply src/util/ajax.ts:283]: Line 283 only calls makeFetchRequest if fetch, Request, AbortController, and AbortController.signal all exist. Line 273 performs no such checks. If there's no entry in config.REGISTERED_PROTOCOLS, makeFetchRequest is used without any further checks.

File: src/util/ajax.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (drwestco) [reply src/util/ajax.ts:351]: If MDN is to be believed, yes. Need to test it though. https://developer.mozilla.org/en-US/docs/Web/API/AbortController/signal

File: src/util/ajax.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (drwestco) [reply src/util/ajax.ts:441]: Doesn't `export const helper ...` on its own accomplish what you need? You shouldn't need separate `export function makeFetchRequest`.

[the remark this answers is not in the document]
Comment (smellyshovel) [reply src/util/ajax.ts:440]: @drwestco `export function xxx` - to be able to use the function in consumer code. `export const helper = {xxx}` - to be able to spy on the function `xxx` in unit tests. I don't understand how to make it different. It's a restriction implied by Jest.

File: src/util/ajax.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (smellyshovel) [reply src/util/ajax.ts]: Tested. Promise resolves without this line instad of being rejected.

File: src/util/ajax.ts (hunk 14 above)
Comment (smellyshovel) [reply src/util/ajax.ts:349]: > Or is the promise rejection intended to bubble up?

Exactly.

File: src/util/ajax.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (smellyshovel) [reply src/util/ajax.ts]: Removed the check.

File: src/util/ajax.ts (hunk 10 above)
Comment (smellyshovel) [reply src/util/ajax.ts:264]: @HarelM Don't get it. There is a function called `getResource` that calls `makeRequest`. I'm not sure why it's done the way it is, but I specifically did not touch it to avoid accidentally breaking something.

File: src/util/ajax.ts (hunk 5 above)
Comment (smellyshovel) [reply src/util/ajax.ts:257]: Done. Now there are regular expressions everywhere. Good point :+1:

File: src/util/ajax.ts (hunk 10 above)
Comment (HarelM) [reply src/util/ajax.ts:264]: This is a call between main thread and the worker, this code can run in both and this is how it's communicating, by calling this method. I don't know how this will work with the promises change...

File: src/util/ajax.ts [hunk 17]
```diff
@@ -365,3 +186,259 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+        if (window.location.protocol === 'blob:') {
+            return window.parent.location.href;
+        } else {
+            return window.location.href;
+        }
+    }
+}
+
+// private functions go below this line. Private functions are dependencies of the public function above and are
+// exported only to be able to be imported in the unit tests
```
[the remark this answers is not in the document]
Comment (HarelM) [reply src/util/ajax.ts:313]: I've added a comment about nise and mocking http/xhr.
I think this is the right direction and not introducing an extra layer to mock out in jest...

File: src/util/ajax.ts (hunk 10 above)
Comment (smellyshovel) [reply src/util/ajax.ts:264]: > This is a call between main thread and the worker, this code can run in both and this is how it's communicating, by calling this method. I don't know how this will work with the promises change...

What is the proper way to test? Is that not covered with all the existing test suits?

Comment (HarelM) [reply src/util/ajax.ts:264]: Maybe with the browser tests, but I would recommend looking at a live vector map in the browser to make sure everything is rendered correctly.

**Comment by zhangyiatmicrosoft** [comment]
ajax.ts has been updated in my PR #2097. If that one gets committed first, I think only one line really needs to be merged <br/>
<pre>requestParameters.headers.accept = 'image/webp,*/*';</pre>
Several test cases were moved as well but I did not actually change them. 
Thanks.

**Comment by smellyshovel** [comment]
> ajax.ts has been updated in my PR #2097. If that one gets committed first, I think only one line really needs to be merged
> 
> requestParameters.headers.accept = 'image/webp,*/*';
> Several test cases were moved as well but I did not actually change them. Thanks.

Generally speaking, I don't see it a good idea to split `getImage` apart from `ajax.ts` file. Logically it belongs there.
On the other hand, it made it easier to resolve the conflicts. But once again, in the end of the day I may still combine them both back together.

File: package.json (hunk 7 above)
Comment (smellyshovel) [reply package.json:34]: @HarelM regarding nise as sinon. As far as I can see, they only allows to deal with XHR API, not Fetch. That's why I introduced the additional dependency. As I remember, `makeFetchRequest` was not covered with tests at all.

Comment (HarelM) [reply package.json:34]: I see, thanks!
Here it is BTW:
https://github.com/sinonjs/nise/issues/7

**Comment by HarelM** [comment]
I would expect to see a queue implantation in one file and maybe the rest in the ajax file as I'm not sure the queue should be part of the ajax, but maybe the ajax file has a bad name and is responsible for too many things... iDK...

File: src/util/ajax.ts (hunk 10 above)
Comment (smellyshovel) [reply src/util/ajax.ts:177]: @HarelM you were right. I only ran manual tests on the 3D map, and they were OK. But the vector one (`debug.html`) was broken.

I've done some research, and it seems to me that that whole main-tread<->worker communication loop for the requests was implemented solely becase back at the time there was no possibility to make a network request from a web worker. But that's not true anymore. I tried to remove all the `isWorker` checks and everything seems to work just fine.

FYI,
Screenshot_20230203_192853

Comment (smellyshovel) [reply src/util/ajax.ts]: BTW, the XHR is also available in web workers, acording to MDN (except for service workers, but I don't know the difference).

Comment (HarelM) [reply src/util/ajax.ts]: If this can be simplified out would be great, but note that the custom protocol which I added can't be used from a web worker because sending a js function is very complicated, so in this case the there's a need to send a message from the worker to the main thread.
But it might be the only use case now...

File: src/source/image_source.test.ts [hunk 18]
```diff
@@ -61,7 +61,9 @@ describe('ImageSource', () => {
         });
         source.onAdd(new StubMap() as any);
         respond();
-        expect(source.image).toBeTruthy();
+        setTimeout(() => {
+            expect(source.image).toBeTruthy();
```
Comment (HarelM) [inline src/source/image_source.test.ts:65]: I think you need to have `done` in the test method signature to indicate that the test is async, otherwise it will pass without really checking the expect value. I might be wrong, but I think this is how it works...

File: src/util/image_request.ts [hunk 19]
```diff
@@ -109,7 +109,7 @@ namespace ImageRequest {
             if (!requestParameters.headers) {
                 requestParameters.headers = {};
             }
-            requestParameters.headers.accept = 'image/webp,*/*';
+            requestParameters.headers['Accept'] = 'image/webp,*/*';
```
Comment (HarelM) [inline src/util/image_request.ts:112]: Is this change really needed?

File: src/util/ajax.ts [hunk 20]
```diff
@@ -278,3 +99,243 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+        TL;DR: for the time being, it's still impossible to completely give up on using the XMLHttpRequest API. But
+        that's a point to reconsider in the (hopefully near) future
+     */
+
+    if (fetch && /:\/\//.test(requestParameters.url) && !(/^https?:|^file:/.test(requestParameters.url))) {
+        // if the url uses some custom protocol. E.g. "custom://..."
+
+        // then check the protocol, and if there exists a custom handler for the protocol, then execute the custom
+        // handler. Otherwise, make the request using the Fetch API
+        const protocol = requestParameters.url.substring(0, requestParameters.url.indexOf('://'));
```
Comment (HarelM) [inline src/util/ajax.ts:175]: If this code executes within a worker the registered protocol method will be null, and nothing will be executed. This generally breaks the `addProtocol` capability. :-(

File: src/util/ajax.ts [hunk 21]
```diff
@@ -278,3 +99,243 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+        response: (async (): Promise<Response<T>> => {
+            const response = await fetch(request);
+
+            if (abortController.signal.aborted) throw new DOMException('aborted', 'AbortError');
+
+            if (response.ok) {
+                let data: T;
+
+                if (requestDataType === 'ArrayBuffer') {
+                    data = await  response.arrayBuffer() as T;
```
Comment (HarelM) [inline src/util/ajax.ts:249]: double space

File: src/util/ajax.ts [hunk 22]
```diff
@@ -278,3 +99,243 @@ export const getVideo = function(urls: Array<string>, callback: Callback<HTMLVid
...
+                } else if (requestDataType === 'JSON') {
+                    data = await response.json() as T;
+                } else {
+                    data = await  response.text() as T;
```
Comment (HarelM) [inline src/util/ajax.ts:253]: double space

**Comment by HarelM** [comment]
WOW, The ajax file looks so much cleaner now! Thanks!!
There's still a need to fix the tests and probably some small fixes, but this looks so much better.
Thanks @smellyshovel!

File: src/util/ajax.ts (hunk 20 above)
Comment (smellyshovel) [reply src/util/ajax.ts:175]: Are you sure, @HarelM? 

Screenshot_20230212_191528

Screenshot_20230212_191508

Comment (smellyshovel) [reply src/util/ajax.ts]: You were right. Fixed by returning the through-actor `postMessage`-communication. The communication model itself is callback-based and that's impossible to change though since `Promise` object are unserializable and thus can't be sent with `postMessage`. So instead the request is performed on the main "thread" and the result is sent back to the worker.

**Comment by smellyshovel** [comment]
@HarelM I played a bit with render tests and can say that to fix some common problems with them it's enough to add the "sleep" operation. But I'm not sure whether it's a good solution (I feel like it's not) and it's still a bit unclear to me (or at least i'm not 100% sure) why is that so.

Do you have any ideas about what might be the reason for them to fail? Any ideas on how to fix?

File: src/util/image_request.ts (hunk 19 above)
Comment (smellyshovel) [reply src/util/image_request.ts]: IDE was complaining... Reverted.

Comment (smellyshovel) [reply src/util/image_request.ts]: See the code hygiene's report.
There are also some failing unit tests because of this file.
Not sure how it happened so that the build passed when this was merged...

**Comment by HarelM** [comment]
The render tests are pretty self contained, they don't do anything "special" in theory.
The following is where a request is sent and also returned, if adding a bit more time there helps, then sure, add it.
https://github.com/maplibre/maplibre-gl-js/blob/ff8e42921f2dcccfffc256190cfbfd0538ebefe9/test/integration/render/run_render_tests.ts#L215

**Comment by HarelM** [comment]
@smellyshovel where are we with this PR?

**Comment by smellyshovel** [comment]
@HarelM 
Still believe that something should be done on that account, but at the moment don't have any time to do so.
I think we can close it now (especially that it's been a long time), and as I feel I'm ready to move on, I'll open another one.

**TL;DR** Sorry, completely forgot and got distracted but other things. Let's close it for now.
